src/hugr/rewrite/replace.rs

```diff
@@ -0,0 +1,591 @@
+//! Implementation of the `Replace` operation.
+
+use std::collections::hash_map::Values;
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::iter::{Chain, Copied, Take};
+use std::slice::Iter;
+
+use itertools::Itertools;
+use thiserror::Error;
+
+use crate::hugr::hugrmut::InsertionResult;
+use crate::hugr::HugrMut;
+use crate::ops::{OpTag, OpTrait};
+use crate::types::EdgeKind;
+use crate::{Direction, Hugr, HugrView, IncomingPort, Node, OutgoingPort};
+
+use super::Rewrite;
+
+/// Specifies how to create a new edge.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct NewEdgeSpec {
+    /// The source of the new edge. For [Replacement::mu_inp] and [Replacement::mu_new], this is in the
+    /// existing Hugr; for edges in [Replacement::mu_out] this is in the [Replacement::replacement]
+    pub src: Node,
+    /// The target of the new edge. For [Replacement::mu_inp], this is in the [Replacement::replacement];
+    /// for edges in [Replacement::mu_out] and [Replacement::mu_new], this is in the existing Hugr.
+    pub tgt: Node,
+    /// The kind of edge to create, and any port specifiers required
+    pub kind: NewEdgeKind,
+}
+
+/// Describes an edge that should be created between two nodes already given
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum NewEdgeKind {
+    /// An [EdgeKind::StateOrder] edge (between DFG nodes only)
+    Order,
+    /// An [EdgeKind::Value] edge (between DFG nodes only)
+    Value {
+        /// The source port
+        src_pos: OutgoingPort,
+        /// The target port
+        tgt_pos: IncomingPort,
+    },
+    /// An [EdgeKind::Static] edge
+    Static {
+        /// The source port
+        src_pos: OutgoingPort,
+        /// The target port
+        tgt_pos: IncomingPort,
+    },
+    /// A [EdgeKind::ControlFlow] edge (between CFG nodes only)
+    ControlFlow {
+        /// Identifies a control-flow output (successor) of the source node.
+        src_pos: OutgoingPort,
+    },
+}
+
+/// Specification of a `Replace` operation
+#[derive(Debug, Clone)] // PartialEq? Eq probably doesn't make sense because of ordering.
```

Comment:
Thoughts here?

Reply:
It is (I had that nondeterministic bug!!) - the new nodes (in the fixed ordering in which `replacement` keeps them) are put into the places of the nodes in `removal` (in that order) - so you can put a new node in as the entry node of a CFG, or an Input/Output, etc.

Also, Hugr itself is only `PartialEq` not `Eq`...but, I guess there is no reason not to make this PartialEq; false negatives is exactly what PartialEq was intended for, even if I suspect we have greater scope for them than, say, `f64` !